query/worker.go

```diff
@@ -0,0 +1,194 @@
+package query
+
+import (
+	"errors"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+var (
+	// ErrQueryTimeout is an error returned if the worker doesn't respond
+	// with a valid response to the request within the timeout.
+	ErrQueryTimeout = errors.New("did not get response before timeout")
+
+	// ErrPeerDisconnected is returned if the worker's peer disconnect
+	// before the query has been answered.
+	ErrPeerDisconnected = errors.New("peer disconnected")
+)
+
+// queryTask is the internal struct that wraps the Query to work on, in
+// addition to some information about the query.
+type queryTask struct {
+	index   int
+	options *queryOptions
+	*Query
+}
+
+// queryTask should satisfy the Task interface in order to be sorted by the
+// workQueue.
+var _ Task = (*queryTask)(nil)
+
+// Index returns the queryTask's index within the work queue.
+func (q *queryTask) Index() int {
+	return q.index
+}
+
+// jobResult is the final result of the worker's handling of the queryTask.
+type jobResult struct {
+	task *queryTask
+	err  error
+}
+
+// worker is responsible for polling work from the work queue, and hand it to
+// the associated peer. It validates incoming responses with the current
+// query's response handler, and poll more work for the peer when it has
+// successfully received a response to the request.
+type worker struct {
+	// peer is the peer this worker will supply with queries, and handle
+	// responses from.
+	peer Peer
+
+	// nextJob is a channel of queries to be distributed, where the worker
+	// will poll new work tasks from.
+	nextJob <-chan *queryTask
+
+	// results is a channel where the worker will deliver the final result
+	// of the last polled task.
+	results chan<- *jobResult
+
+	quit <-chan struct{}
+}
+
+// run starts the worker. The method is blocking, and should be started in a
+// goroutine.
+func (w *worker) run() {
+	// Subscribe to messages from the peer.
+	msgChan, cancel := w.peer.SubscribeRecvMsg()
+	defer cancel()
+
+	for {
+		var job *queryTask
+
+		select {
+		// Poll a new job from the nextJob channel.
+		case job = <-w.nextJob:
+
+		// Ignore any message received while not working on a task.
+		case <-msgChan:
+			continue
+
+		// If the peer disconnected, we can exit immediately, as we
+		// weren't working on a query.
+		case <-w.peer.OnDisconnect():
+			log.Debugf("Peer %v for worker disconnected",
+				w.peer.Addr())
+			return
+
+		case <-w.quit:
+			return
+		}
+
+		// We received a query job, send it to the peer.
+		w.peer.QueueMessageWithEncoding(
+			job.Req, nil, job.options.encoding,
+		)
+
+		// Wait for the correct response to be received from the peer,
+		// or an error happening.
+		var (
+			resp    wire.Message
+			jobErr  error
+			timeout = time.After(job.options.timeout)
```

Comment:
If we want to allow _multiple_ in flight requests (and IMO we should in order to take advantage of batching and pipelining over high latency links), then this timeout value should scale depending on the number of outstanding requests. This addresses the case where we timeout a filter query early due to head of the line blocking in the request queue of the remote peer we're requesting data from. 

Architecture wise, in order to allow multiple in-flight requests, we can keep the design _mostly_ the same, but would then have _multiple_ workers per peer. The scheduler can them inform the worker the timeout value that should be used, as it has a global view of all outstanding requests.

Reply:
Yeah, will revisit when moving to multiple in-flight support